src/FEValues/cell_values.jl

```diff
@@ -35,302 +35,78 @@ function default_geometric_interpolation(::Interpolation{shape}) where {dim, sha
-    dim, T,
-    N_t    <: Vec{dim},
-    dNdx_t <: Tensor{2, dim},
-    dNdξ_t <: Tensor{2, dim},
-}
-    _reinit!(cv, x)
-end
-# vectorized interpolation with vdim != rdim
-function reinit!(cv::CellValues{<:Any, N_t, dNdx_t, dNdξ_t}, x::AbstractVector{Vec{dim,T}}) where {
-    dim, T, vdim,
-    N_t    <: SVector{vdim},
-    dNdx_t <: SMatrix{vdim, dim},
-    dNdξ_t <: SMatrix{vdim, dim},
-}
-    _reinit!(cv, x)
-end
-# reinit! for regular (non-embedded) elements (rdim == sdim)
-function _reinit!(cv::CellValues, x::AbstractVector{Vec{dim,T}}) where {dim, T}
-    n_geom_basefuncs = getngeobasefunctions(cv)
-    n_func_basefuncs = getnbasefunctions(cv)
-    length(x) == n_geom_basefuncs || throw_incompatible_coord_length(length(x), n_geom_basefuncs)
-
-    @inbounds for (i, w) in pairs(getweights(cv.qr))
-        fecv_J = zero(Tensor{2,dim,T})
-        for j in 1:n_geom_basefuncs
-            fecv_J += x[j] ⊗ cv.dMdξ[j, i]
-        end
-        detJ = det(fecv_J)
-        detJ > 0.0 || throw_detJ_not_pos(detJ)
-        cv.detJdV[i] = detJ * w
-        Jinv = inv(fecv_J)
-        for j in 1:n_func_basefuncs
-            # cv.dNdx[j, i] = cv.dNdξ[j, i] ⋅ Jinv
-            cv.dNdx[j, i] = dothelper(cv.dNdξ[j, i], Jinv)
-        end
+struct CellValues{FV, GV, QR, detT<:AbstractVector} <: AbstractCellValues
+    fun_values::FV # FunctionValues
+    geo_values::GV # GeometryValues
+    qr::QR         # QuadratureRule
+    detJdV::detT   # AbstractVector{<:Number}
+end
+function CellValues(::Type{T}, qr::QuadratureRule, ip_fun::Interpolation, ip_geo::VectorizedInterpolation) where T 
+    geo_values = GeometryValues(T, ip_geo.ip, qr, RequiresHessian(ip_fun, ip_geo))
+    fun_values = FunctionValues(T, ip_fun, qr, ip_geo)
+    detJdV = fill(T(NaN), length(getweights(qr)))
+    return CellValues(fun_values, geo_values, qr, detJdV)
+end
+
+CellValues(qr::QuadratureRule, ip::Interpolation, args...) = CellValues(Float64, qr, ip, args...)
+function CellValues(::Type{T}, qr, ip::Interpolation, ip_geo::ScalarInterpolation=default_geometric_interpolation(ip)) where T
+    return CellValues(T, qr, ip, VectorizedInterpolation(ip_geo))
+end
+
+function Base.copy(cv::CellValues)
+    return CellValues(copy(cv.fun_values), copy(cv.geo_values), copy(cv.qr), copy(cv.detJdV))
+end
+
+# Access geometry values
+for op = (:getngeobasefunctions, :geometric_value)
+    eval(quote
+        @propagate_inbounds $op(cv::CellValues, args...) = $op(cv.geo_values, args...)
+    end)
+end
+getdetJdV(cv::CellValues, q_point::Int) = cv.detJdV[q_point]
+
+# Accessors for function values 
+getnbasefunctions(cv::CellValues) = getnbasefunctions(cv.fun_values)
+get_function_interpolation(cv::CellValues) = get_function_interpolation(cv.fun_values)
+get_geometric_interpolation(cv::CellValues) = get_geometric_interpolation(cv.geo_values)
+shape_value_type(cv::CellValues) = shape_value_type(cv.fun_values)
+shape_gradient_type(cv::CellValues) = shape_gradient_type(cv.fun_values)
```

Comment:
Do we really enforce that `shape_value_type` and `shape_gradient_type` have to be the same for function and geometry approximations? I think this is not really true anymore for e.g. complex valued problems.

Reply:
Maybe `function_shape_value_type` to clarify? It is internal usage only anyway, so a bit more verbosity might not hurt and prevent one or another bug. :)

Reply:
This follows `shape_value(::AbstractValues, args...)` and `shape_gradient(::AbstractValues, args...)`, so I think the current state (on master) is acceptable.
But I still see your point - so perhaps we can do that in a separate PR later, I just want to avoid making too many changes here risking to stall this PR.